pkg/reloader/controller.go

```diff
@@ -0,0 +1,222 @@
+// Copyright © 2023 Cisco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reloader
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/bank-vaults/vault-sdk/vault"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	appsinformers "k8s.io/client-go/informers/apps/v1"
+	"k8s.io/client-go/kubernetes"
+	appslisters "k8s.io/client-go/listers/apps/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	DeploymentKind  = "Deployment"
+	DaemonSetKind   = "DaemonSet"
+	StatefulSetKind = "StatefulSet"
+
+	SecretReloadAnnotationName = "alpha.vault.security.banzaicloud.io/reload-on-secret-change"
+	ReloadCountAnnotationName  = "alpha.vault.security.banzaicloud.io/secret-reload-count"
+)
+
+// Controller is the controller implementation for Foo resources
+type Controller struct {
+	kubeClient  kubernetes.Interface
+	vaultClient *vault.Client
+	logger      *logrus.Entry
+
+	deploymentsLister  appslisters.DeploymentLister
+	deploymentsSynced  cache.InformerSynced
+	daemonSetsSynced   cache.InformerSynced
+	daemonSetsLister   appslisters.DaemonSetLister
+	statefulSetsLister appslisters.StatefulSetLister
+	statefulSetsSynced cache.InformerSynced
+
+	// workloadSecrets map[Workload][]string
+	workloadSecrets WorkloadSecretsStore
+	secretVersions  map[string]int
+}
+
+// NewController returns a new sample controller
+func NewController(
+	logger *logrus.Entry,
+	kubeClient kubernetes.Interface,
+	workloadSecrets WorkloadSecretsStore,
```

Comment:
There is no need to pass `WorkloadSecretsStore` to the controller since it holds no logic (it's only a local store). This would allow making the data structure private and only used for the controller logic.